source/common/router/router.cc

```diff
@@ -163,7 +163,17 @@ FilterUtility::finalTimeout(const RouteEntry& route, Http::HeaderMap& request_he
     Http::HeaderEntry* header_expected_timeout_entry =
         request_headers.EnvoyExpectedRequestTimeoutMs();
     if (header_expected_timeout_entry) {
-      trySetGlobalTimeout(header_expected_timeout_entry, timeout);
+      // In case timeout value is out of range,
+      // use *x-envoy-upstream-rq-timeout-ms* instead.
+      if (timeout.global_timeout_ < std::chrono::milliseconds(0)) {
+        Http::HeaderEntry* header_timeout_entry = request_headers.EnvoyUpstreamRequestTimeoutMs();
+        if (trySetGlobalTimeout(header_timeout_entry, timeout)) {
```

Comment:
What do you think about including this logic in `trySetGlobalTimeout`? Seems like it would be good to validate for any global timeout set call, and it would simplify this code too.